### ISSUE TITLE
Incorrect exhaustiveness check on tuple if it's not assigned to a reference before hand py3.12 [1.1.356]

### ISSUE THREAD

**Issue by Vinh-CHUC** [microsoft/pyright#7568]
**Describe the bug**
When pattern matching on a tuple where each component is a Union type, exhaustiveness check (`assert_never`) works properly if the tuple is assigned to a reference before calling `match my_tuple: ..` but not if I write `match (x, y)`.

**Code or Screenshots**

Works:
```python
class A:
    pass
class B:
    pass
class C:
    pass

def g(x: A | B, y: A | B | C):
    arg = (x, y)
    match arg:
        case (A(), A()):
            pass
        case (x, y):
            pass
        case _:
            assert_never(arg)
```

Does not work, only change is inlining the variable:
```python
def g(x: A | B, y: A | B | C):
    
    match (x, y):
        case (A(), A()):
            pass
        case (x, y):
            pass
        case _:
            assert_never((x, y))  # typechecker errors here
```
```shell
 error: Argument of type "tuple[A | B, A | B | C]" cannot be assigned to parameter "arg" of type "Never" in function "assert_never"
    Type "tuple[A | B, A | B | C]" cannot be assigned to type "Never" (reportArgumentType)
```

**Comment by erictraut** [comment]
In general, type narrowing in a `match` statement applies only to the subject expression, not to subexpressions within the subject expression. That is, type narrowing is applied to `args` in your first example and to the expression `(x, y)` in your second example. 

Pyright does have some special-case logic to handle some narrowing in the specific case where the subject expression is a tuple and the subexpressions that make up the tuple are simple identifiers that refer to local variables. You've identified a case that is not currently handled by this special-case logic.

I'll note that other type checkers (including mypy) do not have any special-case handling, so pyright is already doing significantly more here.

Since pyright is working as intended here, I don't consider this a bug. I'm going to change it to an enhancement request and consider it in that light.

One more note: the expression `assert_never((x, y))` will always assert because `(x, y)` always evaluates to a tuple, even if `x` and/or `y` evaluate to `Never`.

**Comment by erictraut** [comment]
This support will be included in the next release.

**Comment by Vinh-CHUC** [comment]
thank you for the extra context, understood

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.357, which I just published.
